File: core/rules/tw_phone_m.py
"""台灣手機號碼（TW_PHONE_M）格式驗證與偵測。"""

import re

# 候選字串：09 開頭 + 8 碼數字，中間可有 0~2 個連字號（如 0912-345-678、0912345678）
# 前後不可緊接數字，避免截斷更長的數字串
_TW_PHONE_M_PATTERN = re.compile(
    r"(?<!\d)09\d{2}-?\d{3}-?\d{3}(?!\d)"
)


def is_valid_tw_phone_m(phone_str: str) -> bool:
    """驗證字串是否為合法台灣手機號碼格式（09 開頭共 10 碼數字）。"""
    if not isinstance(phone_str, str):
        return False

    digits = phone_str.strip().replace("-", "")
    if len(digits) != 10 or not digits.isdigit():
        return False

    return digits.startswith("09")


def detect_tw_phone_m(text: str) -> dict:
    """在文字中找出所有台灣手機號碼，回傳符合 docs/interface.md 格式的偵測結果。"""
    spans = []
    count = 0
    for match in _TW_PHONE_M_PATTERN.finditer(text):
        candidate = match.group()
        if not is_valid_tw_phone_m(candidate):
            continue
        count += 1
        spans.append({
            "start": match.start(),
            "end": match.end(),
            "type": "TW_PHONE_M",
            "text": candidate,
            "confidence": 0.9,
            "source": "rule",
            "replacement": f"[TW_PHONE_M_{count}]",
        })
    return {"text": text, "spans": spans}

File: extension/src/core/tw_phone_m.ts
/**
 * 台灣手機號碼（TW_PHONE_M）格式驗證與偵測。
 * 移植自 Python 版 `core/rules/tw_phone_m.py`，行為必須完全一致。
 */

import type { DetectionResult, Span } from './types';
import { findIter, isAsciiDigits, makeRuleSpan } from './util';

/**
 * 候選字串：09 開頭 + 8 碼數字，中間可有 0~2 個連字號（如 0912-345-678、0912345678）
 * 前後不可緊接數字，避免截斷更長的數字串。
 */
const TW_PHONE_M_PATTERN = /(?<![0-9])09[0-9]{2}-?[0-9]{3}-?[0-9]{3}(?![0-9])/g;

/** 驗證字串是否為合法台灣手機號碼格式（09 開頭共 10 碼數字）。 */
export function isValidTwPhoneM(phoneStr: unknown): boolean {
  if (typeof phoneStr !== 'string') return false;

  const digits = phoneStr.trim().replaceAll('-', '');
  if (digits.length !== 10 || !isAsciiDigits(digits)) return false;

  return digits.startsWith('09');
}

/** 在文字中找出所有台灣手機號碼。 */
export function detectTwPhoneM(text: string): DetectionResult {
  const spans: Span[] = [];
  let count = 0;
  for (const match of findIter(TW_PHONE_M_PATTERN, text)) {
    const candidate = match[0];
    if (!isValidTwPhoneM(candidate)) continue;
    count += 1;
    spans.push(
      makeRuleSpan(
        match.index,
        match.index + candidate.length,
        'TW_PHONE_M',
        candidate,
        0.9,
        count,
      ),
    );
  }
  return { text, spans };
}
